// @genkit-ai/express/lib/index.mjs
import bodyParser from "body-parser";
import cors from "cors";
import { randomUUID } from "crypto";
import express from "express";
import {
  AsyncTaskQueue,
  StreamNotFoundError
} from "genkit/beta";
import {
  getCallableJSON,
  getHttpStatus
} from "genkit/context";
import { logger } from "genkit/logging";
const streamDelimiter = "\n\n";
function expressHandler(action, opts) {
  return async (request, response) => {
    const { stream } = request.query;
    const streamIdHeader = request.headers["x-genkit-stream-id"];
    const streamId = Array.isArray(streamIdHeader) ? streamIdHeader[0] : streamIdHeader;
    if (!request.body) {
      const errMsg = `Error: request.body is undefined. Possible reasons: missing 'content-type: application/json' in request headers or misconfigured JSON middleware ('app.use(express.json()')? `;
      logger.error(errMsg);
      response.status(400).json({ message: errMsg, status: "INVALID ARGUMENT" }).end();
      return;
    }
    const input = request.body.data;
    const init = request.body.init;
    let context;
    try {
      context = await opts?.contextProvider?.({
        method: request.method,
        headers: Object.fromEntries(
          Object.entries(request.headers).filter(([, value]) => value !== void 0).map(([key, value]) => [
            key.toLowerCase(),
            // RFC 9110 5.3: combine repeated field lines with a comma.
            Array.isArray(value) ? value.join(", ") : String(value)
          ])
        ),
        input
      }) || {};
    } catch (e) {
      logger.error(
        `Auth policy failed with error: ${e.message}
${e.stack}`
      );
      response.status(getHttpStatus(e)).json(getCallableJSON(e)).end();
      return;
    }
    const abortController = new AbortController();
    request.on("close", () => {
      abortController.abort();
    });
    request.on("timeout", () => {
      abortController.abort();
    });
    if (request.get("Accept")?.toLowerCase().includes("text/event-stream") || stream === "true") {
      const streamManager = opts?.streamManager;
      if (streamManager && streamId) {
        await subscribeToStream(streamManager, streamId, response);
        return;
      }
      const streamIdToUse = randomUUID();
      const headers = {
        "Content-Type": "text/plain",
        "Transfer-Encoding": "chunked"
      };
      if (streamManager) {
        headers["x-genkit-stream-id"] = streamIdToUse;
      }
      response.writeHead(200, headers);
      runActionWithDurableStreaming(
        action,
        streamManager,
        streamIdToUse,
        input,
        init,
        context,
        response,
        abortController.signal
      );
    } else {
      try {
        const result = await action.run(input, {
          context,
          abortSignal: abortController.signal,
          init
        });
        response.setHeader("x-genkit-trace-id", result.telemetry.traceId);
        response.setHeader("x-genkit-span-id", result.telemetry.spanId);
        response.status(200).json({
          result: result.result
        }).end();
      } catch (e) {
        logger.error(
          `Non-streaming request failed with error: ${e.message}
${e.stack}`
        );
        response.status(getHttpStatus(e)).json(getCallableJSON(e)).end();
      }
    }
  };
}
async function runActionWithDurableStreaming(action, streamManager, streamId, input, init, context, response, abortSignal) {
  let taskQueue;
  let durableStream;
  if (streamManager) {
    taskQueue = new AsyncTaskQueue();
    durableStream = await streamManager.open(streamId);
  }
  try {
    let onChunk = (chunk) => {
      if (response.destroyed) return;
      response.write(
        "data: " + JSON.stringify({ message: chunk }) + streamDelimiter
      );
    };
    if (streamManager) {
      const originalOnChunk = onChunk;
      onChunk = (chunk) => {
        originalOnChunk(chunk);
        taskQueue.enqueue(() => durableStream.write(chunk));
      };
    }
    const result = await action.run(input, {
      onChunk,
      context,
      abortSignal,
      init
    });
    if (streamManager) {
      taskQueue.enqueue(() => durableStream.done(result.result));
      await taskQueue.merge();
    }
    if (!response.destroyed) {
      response.write(
        "data: " + JSON.stringify({ result: result.result }) + streamDelimiter
      );
      response.end();
    }
  } catch (e) {
    if (durableStream) {
      taskQueue.enqueue(() => durableStream.error(e));
      await taskQueue.merge();
    }
    logger.error(
      `Streaming request failed with error: ${e.message}
${e.stack}`
    );
    if (!response.destroyed) {
      response.write(
        `error: ${JSON.stringify({
          error: getCallableJSON(e)
        })}${streamDelimiter}`
      );
      response.end();
    }
  }
}
async function subscribeToStream(streamManager, streamId, response) {
  try {
    await streamManager.subscribe(streamId, {
      onChunk: (chunk) => {
        if (response.destroyed) return;
        response.write(
          "data: " + JSON.stringify({ message: chunk }) + streamDelimiter
        );
      },
      onDone: (output) => {
        if (response.destroyed) return;
        response.write(
          "data: " + JSON.stringify({ result: output }) + streamDelimiter
        );
        response.end();
      },
      onError: (err) => {
        logger.error(
          `Streaming request failed with error: ${err.message}
${err.stack}`
        );
        if (response.destroyed) return;
        response.write(
          `error: ${JSON.stringify({
            error: getCallableJSON(err)
          })}${streamDelimiter}`
        );
        response.end();
      }
    });
  } catch (e) {
    if (response.destroyed) return;
    if (e instanceof StreamNotFoundError) {
      response.status(204).end();
      return;
    }
    if (e.status === "DEADLINE_EXCEEDED") {
      response.write(
        `error: ${JSON.stringify({
          error: getCallableJSON(e)
        })}${streamDelimiter}`
      );
      response.end();
      return;
    }
    throw e;
  }
}
function withContextProvider(flow, context) {
  return {
    flow,
    context
  };
}
function withFlowOptions(flow, options) {
  return {
    flow,
    options
  };
}
function startFlowServer(options) {
  const server = new FlowServer(options);
  server.start();
  return server;
}
class FlowServer {
  /** List of all running servers needed to be cleaned up on process exit. */
  static RUNNING_SERVERS = [];
  /** Options for the flow server configured by the developer. */
  options;
  /** Port the server is actually running on. This may differ from `options.port` if the original was occupied. Null is server is not running. */
  port = null;
  /** Express server instance. Null if server is not running. */
  server = null;
  constructor(options) {
    this.options = {
      ...options
    };
  }
  /**
   * Starts the server and adds it to the list of running servers to clean up on exit.
   */
  async start() {
    const server = express();
    server.use(bodyParser.json(this.options.jsonParserOptions));
    server.use(cors(this.options.cors));
    logger.debug("Running flow server with flow paths:");
    const pathPrefix = this.options.pathPrefix ?? "";
    this.options.flows?.forEach((flow) => {
      if ("flow" in flow) {
        const flowPath = `/${pathPrefix}${"options" in flow && flow.options.path || flow.flow.__action.name}`;
        logger.debug(` - ${flowPath}`);
        const options = "options" in flow ? flow.options : { contextProvider: flow.context };
        server.post(flowPath, expressHandler(flow.flow, options));
      } else {
        const flowPath = `/${pathPrefix}${flow.__action.name}`;
        logger.debug(` - ${flowPath}`);
        server.post(flowPath, expressHandler(flow));
      }
    });
    this.port = this.options?.port || (process.env.PORT ? Number.parseInt(process.env.PORT) : 0) || 3400;
    this.server = server.listen(this.port, () => {
      logger.debug(`Flow server running on http://localhost:${this.port}`);
      FlowServer.RUNNING_SERVERS.push(this);
    });
  }
  /**
   * Stops the server and removes it from the list of running servers to clean up on exit.
   */
  async stop() {
    if (!this.server) {
      return;
    }
    return new Promise((resolve, reject) => {
      this.server.close((err) => {
        if (err) {
          logger.error(
            `Error shutting down flow server on port ${this.port}: ${err}`
          );
          reject(err);
        }
        const index = FlowServer.RUNNING_SERVERS.indexOf(this);
        if (index > -1) {
          FlowServer.RUNNING_SERVERS.splice(index, 1);
        }
        logger.debug(
          `Flow server on port ${this.port} has successfully shut down.`
        );
        this.port = null;
        this.server = null;
        resolve();
      });
    });
  }
  /**
   * Stops all running servers.
   */
  static async stopAll() {
    return Promise.all(
      FlowServer.RUNNING_SERVERS.map((server) => server.stop())
    );
  }
}
export {
  FlowServer,
  expressHandler,
  startFlowServer,
  withContextProvider,
  withFlowOptions
};
